src/cmd/ksh93/tests/whence.sh

```diff
@@ -0,0 +1,63 @@
+# Tests for whence builtin
+
+# ==========
+# -a Displays all uses for each name rather than the first.
+actual=$(whence -a sleep)
+[[ $actual =~ "sleep is a shell builtin" ]] || log_error "whence -a should recognize shell builtins"
+[[ $actual =~ "bin/sleep" ]] || log_error "whence -a should recognize commands in PATH"
+# [[ $actual =~ "sleep is an undefined function" ]] || log_error "whence -a should recognize undefined functions"
+
+
+# ==========
+# -f Do not check for functions.
+function cat {
+    :
+}
+[[ $(whence -f cat) =~ "bin/cat" ]] || log_error "whence -f should ignore functions"
+unset -f cat
+
+# ==========
+# -p Do not check to see if name is a reserved word, a built-in, an alias, or a function. This turns
+#    off the -v option.
+builtin cat
+[[ $(whence -p cat) =~ "bin/cat" ]] || log_error "whence -p should search in PATH"
+builtin -d cat
+
+# ==========
+# -q Quiet mode. Returns 0 if all arguments are built-ins, functions, or are programs found on the
+#    path.
+actual=$(whence -q cat)
+[[ ${#actual} -eq 0 ]] || log_error "whence -q should have empty output"
```

Comment:
Probably worth adding an explicit test of the exit status prior to this line: `[[ $? -eq 0 ]] || ...`. Similarly I would add an explicit test for a command that doesn't exist:

```
actual=$(whence -q no-such-command)
[[ $? -eq 1 ]] || log_error "whence -q of non-existent command should report failure status"
```

Reply:
Added.